Tutorial | Custom Shape Tool in C++ and Python

chanmosq opened the pull request:
Signed-off-by: Chanelle Mosquera <chanmosq@amazon.com>

<!--
    Thanks for your contribution to the Open 3D Engine documentation! Before creating your pull
    request, we ask you to sign off on our submission checklist to ensure that your PR comes through
    quickly. Our reviewers' role is to make sure that all non-trivial submissions follow these best practices.

    By submitting your pull request with DCO signoff, you agree to the Code of Conduct and
    Open 3D Engine documentation and website licensing terms.
-->

## Change summary
Adds a new tutorial for creating custom tools in C++ for O3DE. 

### Submission Checklist:

* [x] **Descriptive active voice** - Do descriptive sentences have a clear *subject* and *action verb*?
* [x] **Answer the question at hand** - Does the documentation answer a *what*, *why*, *how*, or *where* type of question?
* [x] **Consistency** - Does the content consistently follow the [style guide](https://o3de.org/docs/contributing/to-docs/style-guide/)?
* [x] **Help the user** - Does the documentation show the user something *meaningful*?

amzn-leenguy commented:
Hey @chanmosq , could you help me find the page where your content is. I'm new to netlify. thanks

chanmosq commented:
> 
> 
> Hey @chanmosq , could you help me find the page where your content is. I'm new to netlify. thanks

Hey @amzn-leenguy! Here is the link to the preview of the docs on the website: https://deploy-preview-1119--o3deorg.netlify.app/docs/learning-guide/tutorials/custom-tools/shape-example-cpp/#add-a-combo-box

Where did I get the link from? You can preview the website for any PR by doing the following:
1. Go to the bottom of this page where it says **netlify/o3deorg/deploy-preview**, and click **Details.** This will take you to a preview of the website with changes from this PR. 
2. Navigate to the page of the docs you'd like to view. The URL to the docs follows the directory structure of this repo starting at `docs/`. For example, if you want to view the file `content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md`, go to `docs/learning-guide/tutorials/custom-tools/shape-example-cpp/`.

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+First, wrap these UI elements in their own sub-widget, set the layout, and add it to the main widget. 
+
+1. To start, instantiate a `QWidget` called `entityNameWidget` and a `QFormLayout` called `formLayout`.
```
```suggestion
1. To start, instantiate a `QGroupBox` called `entityNameWidget` and a `QFormLayout` called `formLayout`.
```

In the latest code, instead of just a `QWidget` we create a `QGroupBox`, mostly just for stylistic reasons

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Add an input field
+
+Next, create an input field using the `QLineEdit` object called `m_nameInput`. Set the placeholder text in the input field using `setPlaceholderText(...)` and toggle a button to clear the text using `setClearButtonEnabled(...)`.
```
```suggestion
Next, create an input field using the `QLineEdit` object called `m_nameInput`. Set the placeholder text in the input field using `setPlaceholderText(...)` and enable a button (X) to clear the text using `setClearButtonEnabled(...)`.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Add a check box
+
+Then, create a check box using the `QCheckBox` object called `m_addShapeNameSuffix`. You can enable or disable the check box by default using `setDisabled(...)`.
```
I want to change the second sentence, but I'm not sure how. Basically, we add code later in the tutorial that enables/disables the check box dynamically if the user types text into the "entity name" input field. So, we definitely want this to start as disabled, since the "entity name" input field starts off empty.

Maybe we could say something like "You can set the enabled/disabled state on any widget with this API https://doc.qt.io/qt-5/qwidget.html#enabled-prop. This controls whether or not a widget handles mouse/keyboard events, and can change how they are displayed. We will start with this check box in the disabled state, and in the next section will show you how to toggle it at runtime."

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Add a signal listener with a slot handler
+
+Qt uses signals and slots to communicate between objects (refer to [Signals & Slots](https://doc.qt.io/qt-5/signalsandslots.html) in Qt Documentation). You will set up a signal listener such that when the user updates the input field, the check box automatically enables. 
```
```suggestion
Qt uses signals and slots to communicate between objects (refer to [Signals & Slots](https://doc.qt.io/qt-5/signalsandslots.html) in Qt Documentation). You will set up a signal listener such that when the user enters text into the input field, the check box automatically enables. 
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+Qt uses signals and slots to communicate between objects (refer to [Signals & Slots](https://doc.qt.io/qt-5/signalsandslots.html) in Qt Documentation). You will set up a signal listener such that when the user updates the input field, the check box automatically enables. 
+
+In this example, the signal listener uses a slot handler, which is essentially a C++ function that a signal can connect to. 
```
Nice!

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+Again, wrap these UI elements in their own sub-widget, set the layout, and add it to the main widget. 
+
+1. To start, instantiate a `QWidget` called `shapeButtons` and a `QGridLayout` called `gridLayout`.
```
```suggestion
1. To start, instantiate a `QGroupBox` called `shapeButtons` and a `QGridLayout` called `gridLayout`.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+1. To start, instantiate a `QWidget` called `shapeButtons` and a `QGridLayout` called `gridLayout`.
+   
+2. Later, you will query types of shape components, add buttons to this widget, and add signal listeners to the buttons.
```
```suggestion
2. Later, you will query for all types of shape components registered with the engine, add buttons to this widget, and add signal listeners to the buttons.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Query types of Shape components
+
+First, query the types of shape components by specifying a list of provided services. In O3DE, all components have a list of provided services, that you can query from. In this example, Shape components provide the "ShapeService". 
```
```suggestion
First, query the types of shape components registered with the engine. In O3DE, all components have a list of provided services that you can query from. In this example, we are looking for Shape components, which all provide the "ShapeService". 
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Add buttons
+
+Next, you will loop through the list of components, create buttons, and add them to the `gridLayout`. Refer to the following snippet in `ShapeEXampleQWidget.cpp`.
```
```suggestion
Next, you will loop through the list of components, create buttons, and add them to the `gridLayout`. Refer to the following snippet in `ShapeExampleQWidget.cpp`.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+2. Create a button using the QButton object called `shapeButton`. 
+   - Pass in `name` to add the component's name onto the button. 
+   - Set the button minimum height to `40` by using `setMinimumHeight(...)`. 
```
Can remove this line since we aren't using the icons anymore we don't need to set the minimum height.

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+    AZ::TypeId typeId = typeIds[i];
+
+    QPushButton* shapeButton = new QPushButton(QString::fromUtf8(name.c_str()), this);
+    shapeButton->setMinimumHeight(40);
```
Mentioned above, can remove this call to `setMinimumHeight`

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Add a signal listener with a lambda handler
+
+Next, you will create a signal listener such that when the user clicks a button, an entity with the corresponding Shape component will spawn in your scene. 
```
```suggestion
Next, you will create a signal listener such that when the user clicks a button, an entity will be created in your scene and the corresponding Shape component will be added to that entity.
```

I rephrased this slightly to not use the word "spawn", since that has a meaning specific to prefabs.

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
````diff
@@ -0,0 +1,383 @@
+        void CreateEntityWithShapeComponent(const AZ::TypeId& typeId);
+    ```
+
+2. In the for loop that you created earlier, create a connection from the `clicked` signal in the `shapeButton` to the lambda function in `this`. The lambda function calls `CreateEntityWithShapeComponent(...)` and passes in the `typeId`. 
````
```suggestion
2. In the for loop that you created earlier, create a connection from the `clicked` signal in the `shapeButton` to the lambda function in `this`. The lambda function calls `CreateEntityWithShapeComponent(...)` and captures the `typeId` for the shape component being iterated over. 
```

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+weight: 200
+toc: true
+---
+
```
My suggestion here is to add a screenshot of the sample. This will entice and encourage readers to know what is happening on this page.

cgalvan replied:
This should be unblocked now

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+4. Open O3DE Editor for your project.
+
+5. Open the tool by selecting **Tools > Examples**.
+
```
Could we add a screenshot of where this is on the File Menu? Just a nice and simple one. We don't need to screen shot the whole editor. Just the top left section of it with the File Menu open.

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+4. Open O3DE Editor for your project.
+
+5. Open the tool by selecting **Tools > Examples**.
```
Suggestion to change this to: Open the tool by selecting Tools > Examples from the file menu.

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Modules and system components
+
+All Gems have modules and system components that connect the tool to O3DE and allow it to communicate with other systems. The CppToolGem template already contains all of the code needed to run the tool in the O3DE Editor. 
```
Add code pull out for the word **CppToolGem**, in this sentence.

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
````diff
@@ -0,0 +1,383 @@
+```
+
+### Add an input field
+
````
I would love to see screenshots under this headings of what the heading is referring to. E.g. Input field.

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
````diff
@@ -0,0 +1,383 @@
+```
+
+### Add a check box
+
````
I would love to see screenshots under this headings of what the heading is referring to. E.g. Check Box

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+
+## Buttons
+
```
I would love to see screenshots under this headings of what the heading is referring to. E.g. Buttons

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+{{ /note }}
+
+### Add buttons
+
```
I would love to see screenshots under this headings of what the heading is referring to. E.g. Buttons

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+An icon is an image file that's used to represent your tool in the O3DE Editor. You will store the icon using the Qt Resource System and load it from your Gem module. We recommend that your icon adheres to the guidelines in [UI development best practices](https://o3de.org/docs/tools-ui/uidev-component-development-guidelines/#ui-development-best-practices).
+
+### Add an icon
+
```
I would love to see screenshots under this headings of what the heading is referring to. E.g. Icon

amzn-leenguy commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
````diff
@@ -0,0 +1,383 @@
+        ```
+
+    - Call `InitShapeExampleResource()` in the class constructor.
+
````
Add a new call out heading / section here. Similar to what "Icon" is styled as. Maybe you can call it "Download the Sample file"

And have an animated gif of the sample being used in the editor (5-7s loop?)

amzn-leenguy reviewed (changes requested):
added the suggestions and feedback based off our meeting chanelle. thanks!

amzn-leenguy commented:
Here are the additional notes that we spoke about that are not directly related to this page (but i am unsure where to put it).

1. User expected the to have a specific heading on the tutorials landing page. We should add the heading “Extend the Editor” on this page: https://deploy-preview-1119--o3deorg.netlify.app/docs/learning-guide/tutorials/
2. Is Custom Tools or Extend the Editor the right verbiage? Which ever you choose, let me know so I can follow along too.

CC @chanmosq  for visibility

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+
+
+### Communicate with EBusses
```
```suggestion
### Communicate with EBuses
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+### Communicate with EBusses
+
+Now, you will define `CreateEntityWithShapeComponent(...)`, which uses communicates with O3DE EBusses to create a new entity with a component specified by the `typeId` parameter.
```
```suggestion
Now, you will define `CreateEntityWithShapeComponent(...)`, which communicates with O3DE EBuses to create a new entity with a component specified by the `typeId` parameter.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+Now, you will define `CreateEntityWithShapeComponent(...)`, which uses communicates with O3DE EBusses to create a new entity with a component specified by the `typeId` parameter.
+
+1. Use the namespace `AzToolsFramework`. This namespace provides APIs to the necessary Editor EBusses. 
```
```suggestion
1. Use the namespace `AzToolsFramework`. This namespace provides APIs to the necessary Editor EBuses. 
```

I'm not sure if we want to call this out explicitly. I did the `using namespace AzToolsFramework` just to make some of the ebus calls shorter/more readable, since otherwise you'd have to put `AzToolsFramework::` on the beginning of each of them. So maybe we could just say that?  Or nothing at all

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,383 @@
+
+1. Use the namespace `AzToolsFramework`. This namespace provides APIs to the necessary Editor EBusses. 
+
+2. Send a request to create a new entity by calling `EditorRequestBus::BroadcastResult(...)`. Dispatch the `EditorRequests::CreateNewEntity` event, which instantiates an `AZ::EntityId` and stores it in `newEntityId`. 
```
```suggestion
2. Send a request to create a new entity by calling `EditorRequestBus::BroadcastResult(...)`. Dispatch the `EditorRequests::CreateNewEntity` event, which creates a new Entity and returns its `AZ::EntityId`, which is stored in `newEntityId`. 
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
````diff
@@ -0,0 +1,383 @@
+```
+
+
+## Icon
````
This whole section about icons would probably be good to explain, but they don't actually need to add this code. When you create the `ShapeExample` gem from our custom tool gemplate, this code for registering your icon and initializing the resource is already done for you from the template.

chanmosq commented:
> 
> 
> Here are the additional notes that we spoke about that are not directly related to this page (but i am unsure where to put it).
> 
>     1. User expected the to have a specific heading on the tutorials landing page. We should add the heading “Extend the Editor” on this page: https://deploy-preview-1119--o3deorg.netlify.app/docs/learning-guide/tutorials/
> 
>     2. Is Custom Tools or Extend the Editor the right verbiage? Which ever you choose, let me know so I can follow along too.
> 
> 
> CC @chanmosq for visibility

I updated that page and used **Custom tools**.

chanmosq commented:
@amzn-leenguy @cgalvan Thanks for the great feedback!

**My updates:**
- Made changes based on your suggestions.
- Added images where relevant and helpful. (Note that we can't add gifs to the O3DE docs because .gif files are large and we can usually communicate the same information in a single image.)
- Added the Python tutorial here. 

**Remaining task:**
- [ ] Update a few images. I'm currently blocked from this because there's a typo in the Shape Example C++ tool.

amzn-leenguy commented:
@chanmosq can we have the tutorials page, match the order that it is on the left hand side nav? 
https://deploy-preview-1119--o3deorg.netlify.app/docs/learning-guide/tutorials/

Right now it's last, but i'd prefer it in alphabetical

amzn-leenguy commented:
Not sure what is happening here, but it also seems that the links on this page are not working: https://deploy-preview-1119--o3deorg.netlify.app/docs/learning-guide/tutorials/ for custom tools

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,512 @@
+toc: true
+---
+
+In this tutorial, you'll create a custom tool, **Shape Example**, written in C++ that extends **Open 3D Engine (O3DE) Editor**. The Shape Example tool allows you to create entities with a Shape component and configure their component properties. You'll learn how to use the `CppToolGem` template and practice C++ development with O3DE's Tools UI API and other O3DE APIs.
```
```suggestion
In this tutorial, you'll create a custom tool, **Shape Example**, written in C++ that extends the **Open 3D Engine (O3DE) Editor**. The Shape Example tool allows you to create entities with a Shape component and configure their component properties. You'll learn how to use the `CppToolGem` template and practice C++ development with O3DE's Tools UI API and other O3DE APIs.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,512 @@
+
+This tutorial is based off of the **ShapeExample** Gem in [`o3de/sample-code-gems` repository](https://github.com/o3de/sample-code-gems/tree/main/cpp_gems/ShapeExample). You can reference the ShapeExample Gem sample as you follow along this tutorial.
+
+By the end of this tutorial, you'll be able to create your own tools that extends O3DE Editor.
```
```suggestion
By the end of this tutorial, you'll be able to create your own tools that extend the O3DE Editor.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,512 @@
+
+## Create a tool with the `CppToolGem` template
+
+The `CppToolGem` template contains a basic C++ framework to create a dockable widget in O3DE Editor.
```
```suggestion
The `CppToolGem` template contains a basic C++ framework to create a dockable tool (widget) in the O3DE Editor.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
```diff
@@ -0,0 +1,512 @@
+
+### Add a signal listener with a slot handler
+
+Qt uses signals and slots to communicate between objects. Set up a signal listener such that when the user enters text into the input field at runtime, the check box automatically enables. 
```
Might want to have the text `signals and slots` be a link to the Qt docs on signals/slots:  https://doc.qt.io/qt-5/signalsandslots.html

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md:
````diff
@@ -0,0 +1,512 @@
+        void OnNameInputTextChanged(const QString& text);
+    ```
+
+2. Call `QObject::connect(...)` to create a connection between the input field (`m_nameInput`) to the signal (`QLineEdit::TextChanged`) and slot (ShapeExampleWidget::OnNameINputTextChanged`). Refer to the following line of `ShapeExampleWidget.cpp`.
````
```suggestion
2. Call `QObject::connect(...)` to create a connection between the input field (`m_nameInput`) to the signal (`QLineEdit::textChanged`) and slot (ShapeExampleWidget::OnNameInputTextChanged`). Refer to the following line of `ShapeExampleWidget.cpp`.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-py.md:
```diff
@@ -0,0 +1,470 @@
+4. Build the project by using the `o3de` script (refer to [Build a project](https://o3de.org/docs/user-guide/build/configure-and-build/#build-a-project)). Or, use the Project Manager (refer to the **Build** action in the [Project Manager](https://o3de.org/docs/user-guide/project-config/project-manager/)) page.
+
+5. Open O3DE Editor for your project.
+
+6. Open the **Python Scripts** panel by selecting **Tools > Other > Python Scripts** from the file menu. (See A in the following image.)
+
+7. Open `pyshapeexample_dialog` in the Python Scripts panel. (See B in the following image.)
+
```
Discussed this offline, this was a bug due to the fact that our shape example (and `PythonToolGem` template) weren't setting the `QtForPython` gem as a dependency correctly, so it wasn't being bootstrapped properly. This section can be updated to match the C++ version, where it is properly registered in the "Tools -> Examples" menu and has a toolbar icon.

We *should* move these details though down to the debugging section because it is great for quick iteration.

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-py.md:
```diff
@@ -0,0 +1,470 @@
+
+### O3DE and Qt frameworks
+
+O3DE extends the **Qt** framework, so you will use [PySide2](https://pypi.org/project/PySide2/)---a Python module based on [Qt for Python](https://doc.qt.io/qtforpython/)---to create a graphical user interface (GUI). You will also use O3DE's EBuses to communicate with other O3DE interfaces, such as entities and components, and connect them to Qt elements.
```
```suggestion
O3DE extends the **Qt** framework, so you will use [PySide2](https://pypi.org/project/PySide2/)---the Python bindings for Qt [Qt for Python](https://doc.qt.io/qtforpython/)---to create a graphical user interface (GUI). You will also use O3DE's EBuses to communicate with other O3DE interfaces, such as entities and components, and connect them to Qt elements.
```

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-py.md:
```diff
@@ -0,0 +1,470 @@
+
+With Qt for Python, you can create a dialogs which are top-level windows. 
+
+Qt for Python creates a top-level window called a *dialog*. Within a dialog are *widgets*, which are containers for UI elements, and *layouts*, which define how those UI elements are arranged. (In comparison, Qt in C++ creates a main widget instead of a dialog.) Each widget can have its own layout and additional sub-widgets. This nested widget and layout structure allows you to organize groups of ...
```
We might want to discuss changing this to just be a QWidget as well instead of a QDialog for consistency. @HogJonny-AMZN

cgalvan commented on content/docs/learning-guide/tutorials/custom-tools/shape-example-py.md:
```diff
@@ -0,0 +1,470 @@
+
+## Build and debug your tool
+
+[todo]
```
Mentioned above what we want to put in here

cgalvan commented:
Looks great to me! :)  Has gotten very good feedback from external parties as well

amzn-leenguy commented:
> Looks great to me! :) Has gotten very good feedback from external parties as well

Agreed. This has been getting constant 5/5 or 4/5 CSAT ratings from our participants out of 5 people.

The only comment that has resided that you may need to think about is that the title of the Tutorial may need some tweaking. 

I will try and consolidate my findings over the next few days, but some users got mixed up with the term Custom Shape Tool as in a literal custom shape (an obscured Cube or 3d shape). Not a tool for custom shape creation.

This is just a suggestion, but what are your thoughts on: Create a Custom Gem in C++ for Open 3D Engine

feel free to reach out to me if you want to brainstorm more on this.

CC @chanmosq

amzn-leenguy commented on content/docs/learning-guide/tutorials/_index.md:
```diff
@@ -13,6 +13,15 @@ Learn the features, tools, and processes of **Open 3D Engine (O3DE)** with the h
+
+| Tutorial | Description |
+| - | - |
+| [Create a Custom Shape Tool in C++](custom-tools/shape-example-cpp.md) | Create a custom shape tool written in C++ that allows you to quickly create shape objects in your level. Learn about the **CppToolGem** template, the Tools UI API, and C++ development for O3DE. |
```
```suggestion
| [Create a Shape Tool Gem in C++](custom-tools/shape-example-cpp.md) | Create a custom gem in C++ that allows you to quickly create shape objects in your level. Learn about the **CppToolGem** template, the Tools UI API, and C++ development for O3DE. |
```

amzn-leenguy replied:
This may be a bad suggestion, but let me know what you think.

amzn-leenguy commented on content/docs/learning-guide/tutorials/_index.md:
```diff
@@ -13,6 +13,15 @@ Learn the features, tools, and processes of **Open 3D Engine (O3DE)** with the h
+| Tutorial | Description |
+| - | - |
+| [Create a Custom Shape Tool in C++](custom-tools/shape-example-cpp.md) | Create a custom shape tool written in C++ that allows you to quickly create shape objects in your level. Learn about the **CppToolGem** template, the Tools UI API, and C++ development for O3DE. |
+| [Create a Custom Shape Tool in Python](custom-tools/shape-example-py.md) | Create a custom shape tool written in Python that allows you to quickly create shape objects in your level. Learn about the **PythonToolGem** template, the Tools UI API, and Python development for O3DE. |
```
```suggestion
| [Create a Shape Tool Gem in Python](custom-tools/shape-example-py.md) | Create a custom gem written in Python that allows you to quickly create shape objects in your level. Learn about the **PythonToolGem** template, the Tools UI API, and Python development for O3DE. |
```

chanmosq replied:
I decided on the wording "Extend the O3DE Editor by creating a custom tool Gem" and "Create a custom tool Gem". See commit https://github.com/o3de/o3de.org/pull/1119/commits/d918c387c813f1ae29fc299a69f253b3c5a4c716.

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+Define `CreateEntityWithShapeComponent(...)`, which communicates with O3DE EBuses to create a new entity with a component specified by the `typeId` parameter.
+
+1. Send a request to create a new entity by calling `EditorRequestBus::BroadcastResult(...)`. Dispatch the `EditorRequests::CreateNewEntity` event, which creates a new entity and returns its `AZ::EntityId`. The new entity's `AZ::EntityId` is stored in.
```
```suggestion
1. Send a request to create a new entity by calling `EditorRequestBus::BroadcastResult(...)`. Dispatch the `EditorRequests::CreateNewEntity` event, which creates a new entity and returns its `AZ::EntityId`. The new entity's `AZ::EntityId` is stored in `newEntityId`.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+### Add an icon
+
+The following instructions walk you through how to store the icon using the Qt Resource System and load it from your Gem module. Be aware that some of the instructions may already be done by the `PythonToolGem` template.
```
```suggestion
The following instructions walk you through how to store the icon using the Qt Resource System and load it from your Gem module. Be aware that some of the instructions may already be done by the `CppToolGem` template.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+The following instructions walk you through how to store the icon using the Qt Resource System and load it from your Gem module. Be aware that some of the instructions may already be done by the `PythonToolGem` template.
+
+1. Add an image file to the `Code/Source` directory to use as your icon. In this example, the icon is named `toolbar.svg`. We recommend that your icon adheres to the guidelines in [UI development best practices](https://o3de.org/docs/tools-ui/uidev-component-development-guidelines/#ui-development-best-practices).
```
```suggestion
1. Add an image file to the `Code/Source` directory to use as your icon. In this example, the icon is named `toolbar.svg`. We recommend that your icon adheres to the guidelines in [UI development best practices](/docs/tools-ui/uidev-component-development-guidelines/#ui-development-best-practices).
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+1. Add an image file to the `Code/Source` directory to use as your icon. In this example, the icon is named `toolbar.svg`. We recommend that your icon adheres to the guidelines in [UI development best practices](https://o3de.org/docs/tools-ui/uidev-component-development-guidelines/#ui-development-best-practices).
+
+2. Add your icon to PyShapeExample Gem's resources by updating `Code/Source/ShapeExample.qrc` with your new icon's file name.
```
```suggestion
2. Add your icon to MyCppShapeExample Gem's resources by updating `Code/Source/MyCppShapeExample.qrc` with your new icon's file name.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
````diff
@@ -0,0 +1,543 @@
+    
+    ```xml
+    <!DOCTYPE RCC><RCC version="1.0">
+        <qresource prefix="/PyShapeExample">
````
```suggestion
        <qresource prefix="/MyCppShapeExample">
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
````diff
@@ -0,0 +1,543 @@
+    </RCC>
+    ```
+
+3. Register ShapeExample Gem's resources to Qt Resource System by adding the following code in `Code/Source/EditorModule.cpp`.
````
Should this be MyCppShapeExample?

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+3. Register ShapeExample Gem's resources to Qt Resource System by adding the following code in `Code/Source/EditorModule.cpp`.
+
+    - Define the function `InitShapeExampleResource()` and call `Q_INIT_RESOURCE(...)` to register the Qt resrouces listed in `ShapeExample.qrc`.
```
```suggestion
    - Define the function `InitShapeExampleResources()` and call `Q_INIT_RESOURCE(...)` to register the Qt resources listed in `MyCppShapeExample.qrc`.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
````diff
@@ -0,0 +1,543 @@
+    ```cpp
+    void InitShapeExampleResources()
+    {
+        Q_INIT_RESOURCE(ShapeExample);
````
Should this be MyCppShapeExample?
```suggestion
        Q_INIT_RESOURCE(MyCppShapeExample);
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
````diff
@@ -0,0 +1,543 @@
+    }
+    ```
+
+   - Call `InitShapeExampleResource()` in the `EditorModule` class's constructor.
````
```suggestion
   - Call `InitShapeExampleResources()` in the `EditorModule` class's constructor.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+## Build and debug your tool
+
+Build and debug your custom tool using **Visual Studio 2019**.
+
+1. Launch Visual Studio 2019 and open the O3DE solution.
+
+2. Find your Gem in the Solution Explorer. Right-click and select **Build** or **Debug**.
```
I don't think we need steps for Visual Studio, specifically. After all, they could just build again from the command line. They might also be using a different IDE for Linux. Also, there isn't really any information provided here about debugging. Maybe the title could just be named "Build and test your tool". Debugging is sort of implied.

chanmosq replied:
I lean towards keeping the steps in the section in because it shows how to build only the Gem, and not the entire engine, which saves time during development. At the moment, I only tested this on Windows and using Visual Studios.

For now, I can specify that these are for building with Windows, and I can change the title to "Build and test your Gem". (Changing "tool" to "Gem".)

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+2. Find your Gem in the Solution Explorer. Right-click and select **Build** or **Debug**.
+
+After, you can launch Editor and load the Gem.
```
Maybe this should be a step?
```suggestion
Open your project in the Editor.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,543 @@
+
+   - Open the tool in the Editor.
+
+    These steps are explained earlier in this tutorial (refer to [Create a tool with the `CppToolGem` template](#create-a-tool-with-the-cpptoolgem-template)).
```
```suggestion
    These steps are explained earlier in this tutorial. Refer to [Create a tool with the `CppToolGem` template](#create-a-gem-from-the-cpptoolgem-template).
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
```diff
@@ -0,0 +1,508 @@
+toc: true
+---
+
+In this tutorial, you create a custom tool Gem written in Python, called **Shape Example**, that extends the **Open 3D Engine (O3DE) Editor**. The Shape Example tool allows you to create entities with a Shape component and configure their component properties. You'll learn how to use the `PythonToolGem` template and practice Python development with [Qt](https://wiki.qt.io/Main), O3DE Tools UI API ...
```
Many of the same comments in the cpp tutorial apply to the Python tutorial, too.

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,582 @@
+
+1. Create a checkbox by instantiating `QCheckBox` and assigning it to `m_addShapeNameSuffix`.
+
+2. Set the checkbox to the start in the disabled state by using `setDisabled(true)`.
```
```suggestion
2. Set the checkbox to start in the disabled state by using `setDisabled(true)`.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-cpp.md:
```diff
@@ -0,0 +1,582 @@
+
+   - Open the tool in the Editor.
+
+    These steps are explained earlier in this tutorial. Refer to [Create a tool with the `CppToolGem` template](#create-a-gem-from-the-cpptoolgem-template).
```
```suggestion
    These steps are explained earlier in this tutorial. Refer to [Create a Gem from the `CppToolGem` template](#create-a-gem-from-the-cpptoolgem-template).
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
```diff
@@ -0,0 +1,508 @@
+
+This tutorial uses the following directory names and locations in the examples.
+
+- **O3DE engine directory**: `C:/o3de`  
```
Same comment from the C++ tutorial about consistent use of back slashes applies to this Python tutorial.

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
```diff
@@ -0,0 +1,508 @@
+
+All Gems have modules and system component classes written in C++ that connect the tool to O3DE and allow it to communicate with other systems. The `PythonToolGem` template already contains all of the code needed to run the tool in the Editor. You can find these C++ files in `Code/Source`.
+
+For more information on Gem modules and system components, refer to the [Overview of the Open 3D Engine Gem Module system](/docs/user-guide/programming/gems/) page.
```
```suggestion
For more information on Gem modules and system components, refer to the [Overview of the Open 3D Engine Gem Module system](/docs/user-guide/programming/gems/overview/) page.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
```diff
@@ -0,0 +1,508 @@
+
+Before you start the tutorial, ensure that you have the following:
+
+- Set up an O3DE development environment. You can build O3DE as a source engine or a pre-built SDK engine. For more information, refer to [Setting up O3DE from GitHub](/docs/welcome-guide/setup/setup-from-github.md).
```
```suggestion
- Set up an O3DE development environment. For instructions, refer to [Set up Open 3D Engine](/docs/welcome-guide/setup/).
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
````diff
@@ -0,0 +1,508 @@
+        ```cpp
+        void InitShapeExampleResources()
+        {
+            Q_INIT_RESOURCE(ShapeExample);
````
PyShapeExample?
```suggestion
            Q_INIT_RESOURCE(PyShapeExample);
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
```diff
@@ -0,0 +1,508 @@
+
+   - Open the tool in the Editor.
+
+    These steps are explained earlier in this tutorial. Refer to [Create a tool with the `PythonToolGem` template](#create-a-tool-with-the-pythontoolgem-template).
```
```suggestion
    These steps are explained earlier in this tutorial. Refer to [Create a Gem with the `PythonToolGem` template](#create-a-gem-with-the-pythontoolgem-template).
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
```diff
@@ -0,0 +1,508 @@
+
+1. Create a checkbox by instantiating `QCheckBox`. In this example, name it `add_shape_name_suffix`.
+
+2. Set the checkbox to the start in the disabled state by using `setDisabled(True)`.
```
```suggestion
2. Set the checkbox to start in the disabled state by using `setDisabled(True)`.
```

willihay commented on content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md:
```diff
@@ -0,0 +1,508 @@
+
+5. Open Editor for your project.
+
+6. Open the tool by selecting **Tools > Examples > MyPythonShapeExample** from the file menu. (See A in the following image.)
```
```suggestion
6. Open the tool by selecting **Tools > Examples > MyPyShapeExample** from the file menu. (See A in the following image.)
```

chanmosq commented:
@willihay I resolved the last few suggestions, thanks for your attention to detail!
